### PR TITLE
ARROW-13978: [C++] Bump gtest to 1.11 to unbreak builds with recent clang

### PR DESCRIPTION
**Pull request by maflcko**

### CODE REVIEW

**Comment by emkornfield** [comment]
@MarcoFalke it looks like some tests might have failed on windows, I just triggered the rest of CI.

File: cpp/src/arrow/io/compressed_test.cc [hunk 1]
```diff
@@ -273,6 +273,14 @@ TEST(TestSnappyOutputStream, NotImplemented) {
 }
 #endif
 
+#if !defined ARROW_WITH_ZLIB \
+  && !defined ARROW_WITH_BROTLI \
+  && !defined ARROW_WITH_LZ4 \
+  && !defined ARROW_WITH_ZSTD
+GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(CompressedInputStreamTest);
+GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(CompressedOutputStreamTest);
+#endif
```
Comment (pitrou) [inline cpp/src/arrow/io/compressed_test.cc:282]: You'll need to define `GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST` in `arrow/testing/gtest_compat.h` for GTest versions < 1.11.

File: cpp/src/arrow/io/compressed_test.cc [hunk 2]
```diff
@@ -273,6 +273,14 @@ TEST(TestSnappyOutputStream, NotImplemented) {
 }
 #endif
 
+#if !defined ARROW_WITH_ZLIB \
```
Comment (pitrou) [inline cpp/src/arrow/io/compressed_test.cc:276]: I don't think you need to make this conditional.

**Comment by MarcoFalke** [comment]
Thanks, pushed new commit to fix the appveyor errors:

```
16/27 Test #17: arrow-io-compressed-test .........***Failed    0.03 sec
Running main() from C:\projects\arrow\cpp\build-debug\googletest_ep-prefix\src\googletest_ep\googletest\src\gtest_main.cc
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from GoogleTestVerification
[ RUN      ] GoogleTestVerification.UninstantiatedParameterizedTestSuite<CompressedInputStreamTest>
C:/projects/arrow/cpp/src/arrow/io/compressed_test.cc(170): error: Parameterized test suite CompressedInputStreamTest is defined via TEST_P, but never instantiated. None of the test cases will run. Either no INSTANTIATE_TEST_SUITE_P is provided or the only ones provided expand to nothing.
Ideally, TEST_P definitions should only ever be included as part of binaries that intend to use them. (As opposed to, for example, being placed in a library that may be linked in to get other utilities.)
To suppress this error for this test suite, insert the following line (in a non-header) in the namespace it is defined in:
GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(CompressedInputStreamTest);
[  FAILED  ] GoogleTestVerification.UninstantiatedParameterizedTestSuite<CompressedInputStreamTest> (0 ms)
[ RUN      ] GoogleTestVerification.UninstantiatedParameterizedTestSuite<CompressedOutputStreamTest>
C:/projects/arrow/cpp/src/arrow/io/compressed_test.cc(239): error: Parameterized test suite CompressedOutputStreamTest is defined via TEST_P, but never instantiated. None of the test cases will run. Either no INSTANTIATE_TEST_SUITE_P is provided or the only ones provided expand to nothing.
Ideally, TEST_P definitions should only ever be included as part of binaries that intend to use them. (As opposed to, for example, being placed in a library that may be linked in to get other utilities.)
To suppress this error for this test suite, insert the following line (in a non-header) in the namespace it is defined in:
GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(CompressedOutputStreamTest);
[  FAILED  ] GoogleTestVerification.UninstantiatedParameterizedTestSuite<CompressedOutputStreamTest> (0 ms)
[----------] 2 tests from GoogleTestVerification (0 ms total)
[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (1 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] GoogleTestVerification.UninstantiatedParameterizedTestSuite<CompressedInputStreamTest>
[  FAILED  ] GoogleTestVerification.UninstantiatedParameterizedTestSuite<CompressedOutputStreamTest>
 2 FAILED TESTS
```

```
[ RUN      ] GoogleTestVerification.UninstantiatedParameterizedTestSuite<CodecTest>
C:/projects/arrow/cpp/src/arrow/util/compression_test.cc(349): error: Parameterized test suite CodecTest is defined via TEST_P, but never instantiated. None of the test cases will run. Either no INSTANTIATE_TEST_SUITE_P is provided or the only ones provided expand to nothing.
Ideally, TEST_P definitions should only ever be included as part of binaries that intend to use them. (As opposed to, for example, being placed in a library that may be linked in to get other utilities.)
To suppress this error for this test suite, insert the following line (in a non-header) in the namespace it is defined in:
GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(CodecTest);
[  FAILED  ] GoogleTestVerification.UninstantiatedParameterizedTestSuite<CodecTest> (0 ms)
```

**Comment by MarcoFalke** [comment]
Not sure how to fix this:


```
C:/projects/arrow/cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc(1960): error C2065: 'M_PI_2': undeclared identifier

**Comment by pitrou** [comment]
Interesting. It seems that Unity builds may make the `#define _USE_MATH_DEFINES` pointless before `cmath` was already included elsewhere.

**Comment by pitrou** [comment]
I'll try to push a fix.

**Comment by MarcoFalke** [comment]
I also forgot to run clang-format (according to ci):

```diff
--- /arrow/cpp/src/arrow/io/compressed_test.cc
+++ /arrow/cpp/src/arrow/io/compressed_test.cc (after clang format)
@@ -273,10 +273,8 @@
 }
 #endif
 
-#if !defined ARROW_WITH_ZLIB \
-  && !defined ARROW_WITH_BROTLI \
-  && !defined ARROW_WITH_LZ4 \
-  && !defined ARROW_WITH_ZSTD
+#if !defined ARROW_WITH_ZLIB && !defined ARROW_WITH_BROTLI && !defined ARROW_WITH_LZ4 && \
+    !defined ARROW_WITH_ZSTD
 GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(CompressedInputStreamTest);
 GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(CompressedOutputStreamTest);
 #endif

--- /arrow/cpp/src/arrow/util/compression_test.cc
+++ /arrow/cpp/src/arrow/util/compression_test.cc (after clang format)
@@ -584,12 +584,8 @@
   ASSERT_FALSE(result.should_retry);
 }
 
-#if !defined ARROW_WITH_ZLIB \
-  && !defined ARROW_WITH_SNAPPY \
-  && !defined ARROW_WITH_LZ4 \
-  && !defined ARROW_WITH_BROTLI \
-  && !defined ARROW_WITH_BZ2 \
-  && !defined ARROW_WITH_ZSTD
+#if !defined ARROW_WITH_ZLIB && !defined ARROW_WITH_SNAPPY && !defined ARROW_WITH_LZ4 && \
+    !defined ARROW_WITH_BROTLI && !defined ARROW_WITH_BZ2 && !defined ARROW_WITH_ZSTD
 GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(CodecTest);
 #endif

**Comment by pitrou** [comment]
I also did a couple small cleanups along the way. Let's see what CI says.

**Comment by pitrou** [comment]
Hmm, we have a linker error on AppVeyor now.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
Thanks for the contribution @MarcoFalke ! This turned out to be more delicate than expected but it is finally in.

**Comment by MarcoFalke** [comment]
Thanks for picking up. Couldn't have solved this by myself :sweat_smile:

**Pull request closed without merging**